x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportRunAnalyticsAction.java

```diff
@@ -0,0 +1,97 @@
+package org.elasticsearch.xpack.ml.action;
+
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.ActionListenerResponseHandler;
+import org.elasticsearch.action.support.ActionFilters;
+import org.elasticsearch.action.support.HandledTransportAction;
+import org.elasticsearch.action.support.master.AcknowledgedResponse;
+import org.elasticsearch.client.Client;
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.cluster.node.DiscoveryNode;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.common.inject.Inject;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.env.Environment;
+import org.elasticsearch.tasks.Task;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.TransportService;
+import org.elasticsearch.xpack.core.ml.action.RunAnalyticsAction;
+import org.elasticsearch.xpack.core.ml.utils.ExceptionsHelper;
+import org.elasticsearch.xpack.ml.MachineLearning;
+import org.elasticsearch.xpack.ml.analytics.DataFrameDataExtractor;
+import org.elasticsearch.xpack.ml.analytics.DataFrameDataExtractorFactory;
+import org.elasticsearch.xpack.ml.analytics.process.AnalyticsProcessManager;
+
+import java.util.Collections;
+import java.util.Map;
+import java.util.function.Supplier;
+
+public class TransportRunAnalyticsAction extends HandledTransportAction<RunAnalyticsAction.Request, AcknowledgedResponse> {
+
+    private final TransportService transportService;
+    private final ThreadPool threadPool;
+    private final Client client;
+    private final ClusterService clusterService;
+    private final Environment environment;
+    private final AnalyticsProcessManager analyticsProcessManager;
+
+
+    @Inject
+    public TransportRunAnalyticsAction(Settings settings, ThreadPool threadPool, TransportService transportService,
+                                       ActionFilters actionFilters, Client client, ClusterService clusterService, Environment environment,
+                                       AnalyticsProcessManager analyticsProcessManager) {
+        super(settings, RunAnalyticsAction.NAME, transportService, actionFilters,
+            (Supplier<RunAnalyticsAction.Request>) RunAnalyticsAction.Request::new);
+        this.transportService = transportService;
+        this.threadPool = threadPool;
+        this.client = client;
+        this.clusterService = clusterService;
+        this.environment = environment;
+        this.analyticsProcessManager = analyticsProcessManager;
+    }
+
+    @Override
+    protected void doExecute(Task task, RunAnalyticsAction.Request request, ActionListener<AcknowledgedResponse> listener) {
+        DiscoveryNode localNode = clusterService.localNode();
+        if (isMlNode(localNode)) {
```

Comment:
I think this highlights why it would be nice to run the analytics process as a persistent task: then it can be allocated to the ML node with most memory available, taking into account the memory usage of other analytics and anomaly detector processes.

This will require a fair amount of refactoring given that the allocation decision is currently implemented in a nested class inside `TransportOpenJobAction`.